### ISSUE TITLE
cabal haddock --executables fails with "internal error when calculating transitive package dependencies"

### ISSUE THREAD

**Issue by massysett** [haskell/cabal#1919]
When running

```
cabal haddock --executables
```

the command fails with

```
cabal: internal error when calculating transitive package dependencies.
Debug info: []
```

I have tagged my source code tree where this problem exists:

https://github.com/massysett/penny/tree/cabal_error

This is cabal-install version 1.20.0.2.

**Comment by taladar** [comment]
Has any progress been made on this issue? We have some projects with significant code in the modules which only appear in the executable and it would be great if documentation could be generated for those as well. Of course removing the dependency on the local library is not really an option.

**Comment by edsko** [comment]
This problem still occurs with 1.24.0.0.

**Comment by OlivierSohn** [comment]
I'm running into this error, and like @taladar says, removing the dependency on the local library (or convenience library) fixes haddock generation, but I need the dependency nonetheless for projects that depend on the main library (else I have link-time errors).

Has anyone found a workaround for this issue?

Thanks!

**Comment by c-lewis** [comment]
This problem is also present in 2.2.0.1.

**Comment by alexbiehl** [comment]
I think this is fixed with latest Cabal.

Christopher Lewis <notifications@github.com> schrieb am Sa., 27. Okt. 2018,
06:54:

**Comment by int-e** [comment]
FWIW, this bug still exists for the `v1-haddock` command in `cabal-install 2.4.1.0`. It occurs when an executable depends on the library defined in the current package.

**Comment by mietek** [comment]
My `cabal-fail-one` repository is now named [`cabal-issue-1919`](https://github.com/mietek/cabal-issue-1919).

**Comment by int-e** [comment]
Note that `new-haddock` is affected by a *different* issue that also causes executable documentation to not be built in some cases (including @mietek's repository, I think): #5890.

**Comment by madgen** [comment]
I get the same error when I use `cabal haddock` on a package with internal libraries with Cabal 2.4.1.0. Since Stack currently supports up to Cabal 2.4, I imagine this affects some number of people.

Is the only workaround to just not use internal libraries? Does anyone have an idea where to start poking to work towards a patch?

**Comment by brandon-leapyear** [comment]
Seems like this is still an issue? I'm surprised this issue is still open after 7 years. Is there any progress yet? Is this work blocked by another issue, or can it be worked on today?

**Comment by Mikolaj** [comment]
I don't see any label suggesting this is blocked (e.g., on a haddock bug), so I guess any work on this would be very welcome.

I'm surprised this is still failing: I'm running haddock on projects with multiple libraries and it works for me (even if it's rather useless until this haddock work is streamlined:  https://github.com/haskell/haddock/pull/1379). Apparently the pattern of internal library deps matters.

**Comment by brandon-leapyear** [comment]
Here's an up-to-date minimal repro: https://github.com/brandonchinn178/cabal-repro-1919

**Comment by Mikolaj** [comment]
It works for me with development version of cabal (I run `cabal haddock`. Which version and what commandline do you use?

**Comment by brandon-leapyear** [comment]
oh interesting. It fails with `stack haddock`. With debugging enabled, I see it failing when running
```
.../.stack/setup-exe-cache/x86_64-osx/Cabal-simple_mPHDZzAJ_3.2.1.0_ghc-8.10.4 --builddir=.stack-work/dist/x86_64-osx/Cabal-3.2.1.0 haddock --html --hoogle --html-location=../$pkg-$version/ --haddock-option=--hyperlinked-source --haddock-option=--quickjump
```
so maybe it's a GHC bug?

**Comment by Mikolaj** [comment]
Could you make stack build a newer cabal executable? Preferably dev version from branch master? This one looks like 3.2 and it works fine with 3.5 for me and the invocation you show does not see particularly wild, so perhaps this was fixed already?

**Comment by brandon-leapyear** [comment]
Hm so I checked out the `Cabal-v3.4.0.0` tag to build 3.4, and then running `cabal haddock` worked. But then using Cabal 3.4 with Stack still fails
```
$ /Users/bchinn/.stack/setup-exe-cache/x86_64-osx/Cabal-simple_mPHDZzAJ_3.4.0.0_ghc-9.0.1 --builddir=.stack-work/dist/x86_64-osx/Cabal-3.4.0.0 haddock

Preprocessing library for foo-0.1.0.0..
Running Haddock on library for foo-0.1.0.0..
Haddock coverage:
   0% (  0 /  2) in 'Foo'
  Missing documentation for:
    Module header
    doThing (src/Foo.hs:3)
Documentation created:
.stack-work/dist/x86_64-osx/Cabal-3.4.0.0/doc/html/foo/index.html
Preprocessing library 'foo-internal' for foo-0.1.0.0..
Running Haddock on library 'foo-internal' for foo-0.1.0.0..
Cabal-simple_mPHDZzAJ_3.4.0.0_ghc-9.0.1: internal error when calculating
transitive package dependencies.
Debug info: []
```
So it's probably a Stack bug.

I believe this issue can be closed. Mostly because `cabal haddock --executables` seems to not be a thing anymore

**Comment by Mikolaj** [comment]
It would still be nice to report on Stack issue tracker and perhaps improve Cabal to avoid this problem, if it's more on our side.

**Comment by brandon-leapyear** [comment]
It is tracked for Stack: https://github.com/commercialhaskell/stack/issues/4428

But it actually might be a bug in GHC/Cabal, now that I'm digging further. Can you clone my minimal repro again (https://github.com/brandonchinn178/stack-repro-4428) and do the following:
```
ghc-9.0.1 -package Cabal-3.4.0.0 Setup.hs -o setup
./setup configure
./setup haddock
```

**Comment by brandon-leapyear** [comment]
Ok I think I have a better idea of what's going on.

It seems that when calling `cabal haddock`, the `haddock` function (from `Distribution.Simple.Haddock`) is called separately for each component (in the minimal repro, the `foo` library, and the `foo-internal` internal library), so when building the haddocks for `foo-internal`, the `foo` library is registered.

But when manually building `Setup.hs` and calling `haddock` via Cabal-the-library, the `haddock` function is called once for everything, so when building the haddocks for `foo-internal`, the `foo` library doesn't exist.

# Technical details

## cabal haddock

`haddock` is called first with `LocalBuildInfo` containing:
* `configFlags.configArgs = ["lib:foo"]`
* `componentEnabledSpec = OneComponentRequestedSpec (CLibName LMainLibName)`
* `installedPkgs.unitIdIndex` containing:
```
base-4.14.1.0
ghc-prim-0.6.1
integer-gmp-1.0.3.0
rts
```

Then `haddock` is called with `LocalBuildInfo` containing:
* `configFlags.configArgs = ["lib:foo-internal"]`
* `componentEnabledSpec = OneComponentRequestedSpec (CLibName (LSubLibName (UnqualComponentName "foo-internal")))`
* `installedPkgs.unitIdIndex` containing:
```
base-4.14.1.0
foo-0.1.0.0-inplace
ghc-prim-0.6.1
integer-gmp-1.0.3.0
rts
```

## Setup.hs, Cabal-the-library

`haddock` is called once with `LocalBuildInfo` containing:
* `configFlag.configArgs = []`
* `componentEnabledSpec = ComponentRequestedSpec {testsRequested = False, benchmarksRequested = False}`
* `installedPkgs.unitIdIndex` containing:
```
base-4.14.1.0
ghc-prim-0.6.1
integer-gmp-1.0.3.0
rts
```

**Comment by Mikolaj** [comment]
Reproduced. Got `setup: internal error when calculating transitive package dependencies.` What does `./setup configure` even do?

So it looks like cabal problem? The Setup way outdated vs the v2 way?
